File: frontend-v2/src/lib/api/hotCopy.ts
import { api } from "./client";

export interface HotCopyMaterial {
  id: number;
  user_id: number;
  project_id: number | null;
  platform: string;
  source_type: string;
  source_url: string | null;
  account_name: string | null;
  account_home_url: string | null;
  cover_url: string | null;
  title: string;
  original_script: string;
  metrics_json: Record<string, unknown>;
  analysis_json: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
}

export interface CreateManualHotCopyPayload {
  project_id?: number | null;
  platform: "douyin" | "xiaohongshu" | "shipinhao";
  source_url?: string | null;
  account_name?: string | null;
  account_home_url?: string | null;
  cover_url?: string | null;
  title: string;
  original_script: string;
  metrics_json?: Record<string, unknown>;
}

export interface CreateAutoHotCopyPayload {
  project_id?: number | null;
  source_url?: string | null;
  platform?: "douyin" | "xiaohongshu" | "shipinhao";
}

export interface DouyinProfileVideo {
  aweme_id: string;
  video_url: string;
  media_url: string;
  audio_url: string;
  desc: string;
  desc_qualified: boolean;
  create_time: number | null;
  cover_url: string;
  metrics: Record<string, number>;
}

export interface DouyinProfileImportResponse {
  profile: {
    sec_user_id: string;
    nickname: string;
    avatar_url: string;
    signature: string;
    follower_count: number;
    total_favorited: number;
    aweme_count: number;
    source_url: string;
  };
  videos: DouyinProfileVideo[];
  desc_quality: {
    total: number;
    qualified: number;
    qualified_percent: number;
  };
  pagination?: Record<string, unknown>;
}

export interface DouyinProfileTranscriptionResponse {
  aweme_id: string;
  title: string;
  text: string;
  segments: Array<Record<string, unknown>>;
  duration: number | null;
  source_video_oss_key: string;
  source_video_url: string;
  source_video_url_expires_at: number;
}

export interface HotCopyAnalysisResponse {
  material: HotCopyMaterial;
  analysis: Record<string, unknown>;
  generation_record_id: number | null;
}

export interface HotCopyRewritePayload {
  project_id?: number | null;
  rewrite_mode: "light" | "medium" | "strong";
  duration: "30s" | "60s" | "90s";
  conversion_goal: string;
  product?: string | null;
  target_customer?: string | null;
  account_persona?: string | null;
  structure_type?: "talking_head" | "drama" | "mixed" | null;
}

export interface HotCopyRewriteResponse {
  rewrite: {
    id: number;
    material_id: number;
    user_id: number;
    project_id: number | null;
    rewrite_mode: string;
    duration: string;
    conversion_goal: string;
    input_json: Record<string, unknown>;
    output_json: Record<string, unknown>;
    generation_record_id: number | null;
    created_at: string;
  };
  output: Record<string, unknown>;
  generation_record_id: number | null;
}

export async function createManualHotCopyMaterial(payload: CreateManualHotCopyPayload): Promise<HotCopyMaterial> {
  return api.post<HotCopyMaterial>("/api/hot-copy/materials/manual", payload, { timeoutMs: 20000 });
}

export async function createAutoHotCopyMaterial(payload: CreateAutoHotCopyPayload): Promise<HotCopyMaterial> {
  return api.post<HotCopyMaterial>("/api/hot-copy/materials/auto", payload, { timeoutMs: 60000 });
}

export async function importDouyinProfile(sourceUrl: string, count = 30): Promise<DouyinProfileImportResponse> {
  return api.post<DouyinProfileImportResponse>(
    "/api/hot-copy/douyin-profile/import",
    { source_url: sourceUrl, count },
    { timeoutMs: 90000 },
  );
}

export async function transcribeDouyinProfileVideo(payload: {
  aweme_id: string;
  title?: string;
  media_url: string;
  project_id?: number | null;
}): Promise<DouyinProfileTranscriptionResponse> {
  return api.post<DouyinProfileTranscriptionResponse>(
    "/api/hot-copy/douyin-profile/transcribe",
    payload,
    { timeoutMs: 240000 },
  );
}

export async function listHotCopyMaterials(): Promise<HotCopyMaterial[]> {
  return api.get<HotCopyMaterial[]>("/api/hot-copy/materials");
}

export async function analyzeHotCopyMaterial(materialId: number): Promise<HotCopyAnalysisResponse> {
  return api.post<HotCopyAnalysisResponse>(`/api/hot-copy/materials/${materialId}/analyze`, {}, { timeoutMs: 90000 });
}

export async function rewriteHotCopyMaterial(
  materialId: number,
  payload: HotCopyRewritePayload,
): Promise<HotCopyRewriteResponse> {
  return api.post<HotCopyRewriteResponse>(`/api/hot-copy/materials/${materialId}/rewrite`, payload, { timeoutMs: 90000 });
}

export async function searchRedianbaoHotCopy(keyword: string, count = 30): Promise<unknown> {
  return api.post<unknown>("/api/hot-copy/redianbao/search", { keyword, platform: "douyin", count }, { timeoutMs: 20000 });
}

export interface GenerationTaskSubmitResponse {
  task_id: number;
  task_type: string;
  status: string;
  credit_cost: number;
}

export async function generateVideoFromRewrite(rewriteId: number): Promise<GenerationTaskSubmitResponse> {
  return api.post<GenerationTaskSubmitResponse>(`/api/hot-copy/rewrites/${rewriteId}/generate-video`, {}, { timeoutMs: 30000 });
}

export interface SceneTaskInfo {
  task_id: number;
  task_type: string;
  status: string;
  scene_no: number;
  credit_cost: number;
}

export async function generateScenesFromRewrite(rewriteId: number): Promise<{ tasks: SceneTaskInfo[]; total: number }> {
  return api.post<{ tasks: SceneTaskInfo[]; total: number }>(`/api/hot-copy/rewrites/${rewriteId}/generate-scenes`, {}, { timeoutMs: 60000 });
}

File: backend/app/schemas/hot_copy.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


HotCopyPlatform = Literal["douyin", "xiaohongshu", "shipinhao"]
RewriteMode = Literal["light", "medium", "strong"]
RewriteDuration = Literal["30s", "60s", "90s"]


class HotCopyMaterialManualCreate(BaseModel):
    project_id: int | None = Field(default=None, gt=0)
    platform: HotCopyPlatform = "douyin"
    source_url: str | None = Field(default=None, max_length=1000)
    account_name: str | None = Field(default=None, max_length=120)
    account_home_url: str | None = Field(default=None, max_length=1000)
    cover_url: str | None = Field(default=None, max_length=1000)
    title: str = Field(min_length=1, max_length=240)
    original_script: str = Field(min_length=1, max_length=12000)
    metrics_json: dict[str, Any] = Field(default_factory=dict)

    @field_validator("title", "original_script", mode="before")
    @classmethod
    def strip_required_string(cls, value: str) -> str:
        if isinstance(value, str):
            return value.strip()
        return value

    @field_validator("source_url", "account_name", "account_home_url", "cover_url", mode="before")
    @classmethod
    def strip_optional_string(cls, value: str | None) -> str | None:
        if value is None:
            return None
        if isinstance(value, str):
            cleaned = value.strip()
            return cleaned or None
        return value


class HotCopyMaterialAutoCreate(BaseModel):
    project_id: int | None = Field(default=None, gt=0)
    source_url: str | None = Field(default=None, max_length=1000)
    platform: HotCopyPlatform = "douyin"

    @field_validator("source_url", mode="before")
    @classmethod
    def strip_optional_string(cls, value: str | None) -> str | None:
        if value is None:
            return None
        if isinstance(value, str):
            cleaned = value.strip()
            return cleaned or None
        return value


class DouyinProfileImportRequest(BaseModel):
    source_url: str = Field(min_length=1, max_length=1000)
    count: int = Field(default=30, ge=1, le=50)

    @field_validator("source_url", mode="before")
    @classmethod
    def strip_source_url(cls, value: str) -> str:
        if isinstance(value, str):
            return value.strip()
        return value


class DouyinProfileTranscribeRequest(BaseModel):
    aweme_id: str = Field(min_length=1, max_length=80)
    media_url: str = Field(min_length=1, max_length=5000)
    title: str = Field(default="", max_length=240)
    project_id: int | None = Field(default=None, gt=0)

    @field_validator("aweme_id", "media_url", "title", mode="before")
    @classmethod
    def strip_text(cls, value: str) -> str:
        if isinstance(value, str):
            return value.strip()
        return value


class HotCopyMaterialRead(BaseModel):
    id: int
    user_id: int
    project_id: int | None = None
    platform: str
    source_type: str
    source_url: str | None = None
    account_name: str | None = None
    account_home_url: str | None = None
    cover_url: str | None = None
    title: str
    original_script: str
    metrics_json: dict[str, Any] = Field(default_factory=dict)
    analysis_json: dict[str, Any] | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class HotCopyAnalysisResponse(BaseModel):
    material: HotCopyMaterialRead
    analysis: dict[str, Any]
    generation_record_id: int | None = None


class HotCopyRewriteRequest(BaseModel):
    project_id: int | None = Field(default=None, gt=0)
    rewrite_mode: RewriteMode = "medium"
    duration: RewriteDuration = "60s"
    conversion_goal: str = Field(default="私信获客", min_length=1, max_length=80)
    product: str | None = Field(default=None, max_length=200)
    target_customer: str | None = Field(default=None, max_length=200)
    account_persona: str | None = Field(default=None, max_length=200)
    structure_type: Literal["talking_head", "drama", "mixed"] | None = Field(default=None)

    @field_validator("structure_type", mode="before")
    @classmethod
    def normalize_structure_type(cls, value: str | None) -> str | None:
        if value is None:
            return None
        cleaned = str(value).strip().lower()
        if cleaned in {"talking_head", "drama", "mixed"}:
            return cleaned
        return None

    @field_validator("conversion_goal", mode="before")
    @classmethod
    def strip_required_string(cls, value: str) -> str:
        if isinstance(value, str):
            return value.strip()
        return value

    @field_validator("product", "target_customer", "account_persona", mode="before")
    @classmethod
    def strip_optional_string(cls, value: str | None) -> str | None:
        if value is None:
            return None
        if isinstance(value, str):
            cleaned = value.strip()
            return cleaned or None
        return value


class HotCopyRewriteRead(BaseModel):
    id: int
    material_id: int
    user_id: int
    project_id: int | None = None
    rewrite_mode: str
    duration: str
    conversion_goal: str
    input_json: dict[str, Any] = Field(default_factory=dict)
    output_json: dict[str, Any] = Field(default_factory=dict)
    generation_record_id: int | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class HotCopyRewriteResponse(BaseModel):
    rewrite: HotCopyRewriteRead
    output: dict[str, Any]
    generation_record_id: int | None = None


class HotCopyRedianbaoSearchRequest(BaseModel):
    keyword: str = Field(min_length=1, max_length=200)
    platform: HotCopyPlatform = "douyin"
    count: int = Field(default=30, ge=30, le=100)

    @field_validator("keyword", mode="before")
    @classmethod
    def strip_keyword(cls, value: str) -> str:
        if isinstance(value, str):
            return value.strip()
        return value
